### PR TITLE
Add expand_tuple utility

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

A utility function that helps expanding a tuple in a parameter pack. Simplifies some existing code and also helps with code in #2077 and upcoming PRs.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by moxcodes** [comment]
how does this differ from `std::apply`? https://en.cppreference.com/w/cpp/utility/apply
(though, I think we're still not all up to cpp17 with our supported compilers, so the change request here would be to call it `cpp17::apply` and change it to the example implementation at cppreference with the forwards)

**Comment by nilsvu** [comment]
@moxcodes looks pretty much the same! It's only available from cpp17 though, so I moved the function to `Utilities/Tuple.hpp` and into the `cpp17` namespace

File: src/Utilities/Tuple.hpp [hunk 1]
```diff
@@ -174,3 +174,37 @@ constexpr inline void tuple_transform(
...
+ * Here is an example how to use the function:
+ *
+ * \snippet Test_Tuple.cpp expand_tuple_example
+ *
+ * This is the function being called in the above example:
+ *
+ * \snippet Test_Tuple.cpp expand_tuple_example_function
+ */
+template <typename Invokable, typename... Args>
+auto apply(Invokable&& invokable, const std::tuple<Args...>& args) {
```
Comment (moxcodes) [inline src/Utilities/Tuple.hpp]: I think it would be better to take the tuple by rvalue reference, and to `std::forward` when passing to the impl, and do the same rvalue reference and forwarding in the implementation.
I see the "possible implementation" on cppreference uses a cpp20 feature `invoke`... however, the main complexities of `invoke` involve handling member functions and member objects, so I think the best way to go about this is to just assume for the moment that we're not applying member function pointers and use the simple version in the `apply_impl`:
```
return std::forward<Invokable>(invokable)(std::forward<Args>(args)...)
```
and make the return type of the functions `decltype(auto)` and the functions `constexpr` in case the invokable is sufficiently simple to be brought in at compile-time.

Comment (nilsvu) [reply src/Utilities/Tuple.hpp]: agreed, I pushed a fixup a few minutes ago that does pretty much that :D

Comment (nilsvu) [reply src/Utilities/Tuple.hpp]: one question though: do I need the `std::forward<Invokable>(invokable)` at the beginning? I'm not quite sure what that's for

Comment (moxcodes) [reply src/Utilities/Tuple.hpp]: I think it is best to do that for funny corner cases. My understanding is that it ensures appropriate forwarding of the invokable as reference. So for instance, if passing a non-const lvalue reference to a class with a call operator that alters the class member variables, I think it would appropriately ensure that the class members get altered; and call the right version of overloaded call operators when deciding between const and non-const versions according to whether the invokable was passed as const or non-const reference.

**Review by moxcodes — COMMENTED** [review]
looks good, thanks for making those changes. Go ahead and squash.

**Review by moxcodes — DISMISSED** [review]

File: src/Utilities/Tuple.hpp [hunk 2]
```diff
@@ -174,3 +174,36 @@ constexpr inline void tuple_transform(
       tuple, std::forward<N_aryOp>(op),
       std::make_index_sequence<sizeof...(Elements)>{}, args...);
 }
+
+namespace tuple_impl_detail {
```
Comment (nilsdeppe) [inline src/Utilities/Tuple.hpp]: why not moving this into the `cpp17` namespace?

File: tests/Unit/Utilities/Test_Tuple.cpp [hunk 3]
```diff
@@ -249,3 +249,23 @@ SPECTRE_TEST_CASE("Unit.Utilities.tuple_transform", "[Utilities][Unit]") {
...
+double test_function(const int first_arg, const double second_arg) {
+  return static_cast<double>(first_arg) + second_arg;
+}
+/// [expand_tuple_example_function]
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Utilities.ExpandTuple", "[Unit][Utilities]") {
+  /// [expand_tuple_example]
+  const double extra_factor = 3.;
+  const double result = cpp17::apply(
```
Comment (nilsdeppe) [inline tests/Unit/Utilities/Test_Tuple.cpp:264]: Can you add a test with a tuple of 1 element and 0 elements?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 4]
```diff
@@ -103,12 +60,23 @@ void verify_solution(
       solution.variables(inertial_coords, all_fields{}));
 
   // Apply operator to solution fields
-  auto fluxes = detail::ExpandTuplesImpl<System>::first_order_fluxes(
-      solution_fields, fluxes_computer, fluxes_args);
+  auto fluxes = cpp17::apply(
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:63]: Please move these changes into a separate commit.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm fine with you squashing immediately

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsvu** [comment]
@moxcodes could you take another quick look please? Thanks for reviewing @nilsdeppe and @moxcodes!

**Review by moxcodes — APPROVED** [review]
